src/impy/models/sophia.py

```diff
@@ -142,22 +142,28 @@ def _set_event_kinematics(self, event_kinematics):
         k = event_kinematics
 
         if k.p1pdg != 22:
-            info(0, "The first particle must be a photon, but particle pdg = ", k.p1pdg)
-            raise Exception("Input error")
+            raise ValueError(
+                "Sophia accepts only 'gamma' as a projectile "
+                + "but received pdg_id = {0}".format(k.p1pdg)
+            )
 
         if k.p2pdg not in [2212, 2112]:
-            info(
-                0,
-                "The second particle must be a proton or neutron, but particle pdg = ",
-                k.p2pdg,
+            raise ValueError(
+                "Sophia accepts only 'proton' or 'neutron' as a target, "
+                + "but received pdg_id = {0}".format(k.p2pdg)
+            )
+
+        if k.p2_is_nucleus:
+            raise ValueError(
+                "Sophia accepts only 'proton' or 'neutron' as a target, "
+                + "but received nucleus = ({0}, {1})".format(k.A2, k.Z2)
             )
-            raise Exception("Input error")
 
         self.nucleon_code_number = self.lib.icon_pdg_sib(k.p2pdg)
         self.lib.initial(
             self.nucleon_code_number
         )  # setting parameters for cross-section
-        self.energy_of_nucleon = k.pmass2
+        self.energy_of_nucleon = np.float32(k.pmass2)  # fix roundoff error
```

Comment:
I don't understand, can you elaborate what is the problem? np.float32 has lower resolution than the standard Python `float` which is `np.float64`?

Reply:
This workaround is required because there is a subtraction of a small number in sophia and roundoff errors matter. Sophia initializes a double precision mass array with single precision literals, i.e. `/0.9378/' instead of `/0.9378D0/`. So it compares a true DP representation of a number with that of a DBLE cast. To not change stuff in sophia inconsistently we just pass the argument in a way that is doing the same. I preferred this solution compared to hacking sophia.

Reply:
Ok, that makes sense! Could you please add a more elaborate comment to explain to future readers of the code? This is a rather subtle issue and with your longer explanation I can actually understand why it has to be like this. You basically want to make sure that the calculation inside Sophia and in Python give bit-identical results, even though this change reduces the accuracy of self.energy_of_nucleon, right?